### PR TITLE
MG-2380 - Pass env files to vault script

### LINKED ISSUE #2380: Feature: Pass env files to vault scripts
### Is your feature request related to a problem? Please describe.

I'm trying to use Vault Scripts with .env file in different path. 
Present vault script doesn't support , it is fixed to docker/.env

### Describe the feature you are requesting, as well as the possible use case(s) for it.

Passing env file path as cli args to vault scripts 

### Indicate the importance of this feature to you.

Must-have

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by JeffMboya**
# What type of PR is this?

This is a feature because it adds the functionality to pass the `.env` file path as CLI arguments to Vault scripts.

## What does this do?

This PR modifies the Vault scripts to support passing the `.env` file path as a command-line argument. It updates the `readDotEnv` function to use the provided `.env` file path and handles the default path if no argument is provided.

## Which issue(s) does this PR fix/relate to?

- Related Issue https://github.com/absmach/devops/issues/135
- Resolves #2380

## Have you included tests for your changes?

No
## Did you document any new/modified feature?

Yes, I have updated the documentation for the new feature.

### Notes

This update ensures that the Vault scripts are more flexible and can work with `.env` files located in different paths.

### CODE REVIEW

File: docker/addons/vault/vault_copy_certs.sh [hunk 1]
```diff
@@ -9,9 +9,21 @@ export MAGISTRALA_DIR=$scriptdir/../../../
...
+# Parse command line arguments
+while [[ "$#" -gt 0 ]]; do
+    case $1 in
+        --env-file) env_file="$2"; shift ;;
+        *) echo "Unknown parameter passed: $1"; exit 1 ;;
+    esac
+    shift
+done
+
 readDotEnv() {
```
Comment (arvindh123) [inline docker/addons/vault/vault_copy_certs.sh]: So something like this we can try 
```suggestion

# Default .env file path
env_file="docker/.env"

# Parse command line arguments
while [[ "$#" -gt 0 ]]; do
    case $1 in
        --env-file) env_file="$2"; shift ;;
        *) echo "Unknown parameter passed: $1"; exit 1 ;;
    esac
    shift
done

certs_copy_path="docker/ssl/certs/"
while [[ "$#" -gt 0 ]]; do
    case $1 in
        --dest-path) certs_copy_path="$2"; shift ;;
        *) echo "Unknown parameter passed: $1"; exit 1 ;;
    esac
    shift
done


readDotEnv() {
    set -o allexport
    source "$env_file"
    set +o allexport
}

readDotEnv

server_name="localhost"

# Check if MG_NGINX_SERVER_NAME is set or not empty
if [ -n "${MG_NGINX_SERVER_NAME:-}" ]; then
    server_name="$MG_NGINX_SERVER_NAME"
fi

echo "Copying certificate files"

if [ -e "data/${server_name}.crt" ]; then
    cp -v data/${server_name}.crt      ${certs_copy_path}/magistrala-server.crt
else
    echo "${server_name}.crt file not available"
fi

if [ -e "data/${server_name}.key" ]; then
    cp -v data/${server_name}.key      ${certs_copy_path}/magistrala-server.key
else
    echo "${server_name}.key file not available"
fi

if [ -e "data/${MG_VAULT_PKI_INT_FILE_NAME}.key" ]; then
    cp -v data/${MG_VAULT_PKI_INT_FILE_NAME}.key    ${certs_copy_path}/ca.key
else
    echo "data/${MG_VAULT_PKI_INT_FILE_NAME}.key file not available"
fi

if [ -e "data/${MG_VAULT_PKI_INT_FILE_NAME}_bundle.crt" ]; then
    cp -v data/${MG_VAULT_PKI_INT_FILE_NAME}_bundle.crt     ${certs_copy_path}/ca.crt
else
    echo "data/${MG_VAULT_PKI_INT_FILE_NAME}_bundle.crt file not available"
fi

exit 0

```

**Review by arvindh123 — CHANGES_REQUESTED** [review]

File: docker/addons/vault/vault_copy_certs.sh (hunk 1 above)
Comment (arvindh123) [reply docker/addons/vault/vault_copy_certs.sh]: [> So something like this we can try](https://github.com/absmach/magistrala/pull/2387#discussion_r1726947534)

After implementing this please  check it's working  with https://github.com/absmach/amdm/pull/88

Comment (arvindh123) [reply docker/addons/vault/vault_copy_certs.sh]: @JeffMboya Please move all the scripts to a directory

**Review by arvindh123 — CHANGES_REQUESTED** [review]

File: docker/addons/vault/vault_copy_certs.sh (hunk 1 above)
Comment (JeffMboya) [reply docker/addons/vault/vault_copy_certs.sh]: @JeffMboya 
@arvindh123 A quick clarification: move the scripts to a single directory inside the vault folder i.e `docker/addons/vault/scripts` or outside the vault folder i.e magistrala/scripts? https://github.com/absmach/magistrala/pull/2387#discussion_r1727039272


@Arvindh
Good question,
I think we can move to `docker/addons/vault/scripts` for now
What do you think?

@JeffMboya
I'm thinking of using `docker/addons/vault/scripts` because it keeps the scripts together with other Vault files.

@arvindh123
okay then lets go with `docker/addons/vault/scripts`

Comment (JeffMboya) [reply docker/addons/vault/vault_copy_certs.sh]: All the vault scripts have been moved to `docker/addons/vault/scripts`

File: docker/addons/vault/scripts/vault_init.sh [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+source vault_cmd.sh
+
+readDotEnv
+
+# Create the data directory inside docker/addons/vault/scripts
+mkdir -p data
+
+vault operator init -address=$MG_VAULT_ADDR 2>&1 | tee >(sed -r 's/\x1b\[[0-9;]*m//g' > data/secrets)
```
Comment (arvindh123) [inline docker/addons/vault/scripts/vault_init.sh]: Add new line

**Review by arvindh123 — CHANGES_REQUESTED** [review]

File: docker/addons/vault/docker-compose.yml [hunk 3]
```diff
@@ -26,7 +26,7 @@ services:
       - magistrala-vault-volume:/vault/file
       - magistrala-vault-volume:/vault/logs
       - ./config.hcl:/vault/config/config.hcl
-      - ./entrypoint.sh:/entrypoint.sh
+      - ./scripts/entrypoint.sh:/entrypoint.sh
```
Comment (arvindh123) [inline docker/addons/vault/docker-compose.yml]: @JeffMboya This can be in `docker/addons/vault`, becuase it is script for docker , and not for setup

**Review by arvindh123 — CHANGES_REQUESTED** [review]

**Review by arvindh123 — CHANGES_REQUESTED** [review]
Move docker/addons/vault/.gitignore to scripts directory

File: docker/addons/vault/scripts/vault_copy_certs.sh [hunk 4]
```diff
@@ -5,13 +5,27 @@
...
+certs_copy_path="../../../../docker/ssl/certs/"
+
+# Parse command line arguments
+while [[ "$#" -gt 0 ]]; do
+    case $1 in
+        --env-file) env_file="$2"; shift ;;
+        *) echo "Unknown parameter passed: $1"; exit 1 ;;
+    esac
+    shift
+done
```
Comment (arvindh123) [inline docker/addons/vault/scripts/vault_copy_certs.sh:43]: Please add options to pass `certs_copy_path`

**Review by arvindh123 — CHANGES_REQUESTED** [review]

**Review by arvindh123 — CHANGES_REQUESTED** [review]
For this bash command from root of repo 
```bash 
docker/addons/vault/scripts/vault_init.sh --env-file docker/.env
````
I get below error
```bash
docker/addons/vault/scripts/vault_init.sh: line 25: docker/.env: No such file or directory
```

**Comment by JeffMboya** [comment]
> For this bash command from root of repo
> 
> ```shell
> docker/addons/vault/scripts/vault_init.sh --env-file docker/.env
> ```
> 
> I get below error
> 
> ```shell
> docker/addons/vault/scripts/vault_init.sh: line 25: docker/.env: No such file or directory
> ```

@arvindh123 this has been resolved

File: docker/addons/vault/scripts/vault_init.sh [hunk 5]
```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/bash
+# Copyright (c) Abstract Machines
+# SPDX-License-Identifier: Apache-2.0
+
+set -euo pipefail
+
+scriptdir="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd)"
+repo_root="$(realpath "$scriptdir/../../../../")"
+env_file="$repo_root/docker/.env"
```
Comment (arvindh123) [inline docker/addons/vault/scripts/vault_init.sh]: Let the `env_file` path be relative `docker/.env` and remove the `repo_root`

File: docker/addons/vault/scripts/vault_set_pki.sh [hunk 6]
```diff
@@ -58,9 +77,9 @@ vaultGenerateRootCACertificate() {
         province="\"$MG_VAULT_PKI_CA_ST\"" \
         street_address="\"$MG_VAULT_PKI_CA_ADDR\"" \
         postal_code="\"$MG_VAULT_PKI_CA_PO\"" \
-        ttl=87600h | tee >(jq -r .data.certificate >data/${MG_VAULT_PKI_FILE_NAME}_ca.crt) \
-                         >(jq -r .data.issuing_ca  >data/${MG_VAULT_PKI_FILE_NAME}_issuing_ca.crt) \
-                         >(jq -r .data.private_key >data/${MG_VAULT_PKI_FILE_NAME}_ca.key)
+        ttl=87600h | tee >(jq -r .data.certificate >"$repo_root/docker/addons/vault/scripts/data/${MG_VAULT_PKI_FILE_NAME}_ca.crt") \
```
Comment (arvindh123) [inline docker/addons/vault/scripts/vault_set_pki.sh]: Instead of `$repo_root/docker/addons/vault/scripts/data` use `$scriptdir/data`
And remove `$repo_root`

**Review by arvindh123 — CHANGES_REQUESTED** [review]

File: docker/addons/vault/scripts/vault_copy_env.sh [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+scriptdir="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd)"
+
+# default env file path
+env_file="docker/.env"
+
+while [[ "$#" -gt 0 ]]; do
+    case $1 in
+        --env-file)
+            if [[ "$2" = /* ]]; then
```
Comment (arvindh123) [inline docker/addons/vault/scripts/vault_copy_env.sh]: @JeffMboya What does it does ? 
```bash
if [[ "$2" = /* ]]; then
```
If i pass /* to --env-file 
```bash
docker/addons/vault/scripts/vault_copy_env.sh --env-file /*
```
then getting a wired error 
```
Unknown parameter passed: /bin
```

**Review by arvindh123 — CHANGES_REQUESTED** [review]

File: docker/addons/vault/scripts/vault_copy_env.sh (hunk 7 above)
Comment (arvindh123) [reply docker/addons/vault/scripts/vault_copy_env.sh]: Simply we can throw error , If `$2` is empty.

Comment (JeffMboya) [reply docker/addons/vault/scripts/vault_copy_env.sh]: @arvindh123 This checks if the second argument (`$2`) is an absolute path, which in Unix/Linux means it starts with a `/`. If it is an absolute path, the script simply assigns it to the `env_file` variable as it is.

Comment (JeffMboya) [reply docker/addons/vault/scripts/vault_copy_env.sh]: The error happens because `/*` is a wildcard pattern that matches all directories and files at the root level such as `/bin`, `/etc`, `/lib`, etc., and the script is picking up `/bin` as one of those matches. The script then attempts to process `/bin` as the value of `--env-file`, which it doesn't expect, leading to the `"Unknown parameter passed: /bin"` error.

Comment (arvindh123) [reply docker/addons/vault/scripts/vault_copy_env.sh]: Why do we need this logic ? Just trying to understand you code 
Particularly why we need to convert relative path to Absolute Path ? 
```bash
            if [[ "$2" = /* ]]; then
                env_file="$2"
            else
                env_file="$(realpath -m "$PWD/$2")"
            fi
            shift
            ;;
```

Comment (JeffMboya) [reply docker/addons/vault/scripts/vault_copy_env.sh]: > Simply we can throw error , If `$2` is empty.

@arvindh123 an error is thrown when `$2` is empty:
```sh
$ magistrala/docker/addons/vault/scripts/vault_copy_env.sh --env-file   
magistrala/docker/addons/vault/scripts/vault_copy_env.sh: line 15: $2: unbound variable
```

Comment (JeffMboya) [reply docker/addons/vault/scripts/vault_copy_env.sh]: > Why do we need this logic ? Just trying to understand you code Particularly why we need to convert relative path to Absolute Path ?
> 
> ```shell
>             if [[ "$2" = /* ]]; then
>                 env_file="$2"
>             else
>                 env_file="$(realpath -m "$PWD/$2")"
>             fi
>             shift
>             ;;
> ```

The logic is included in the script to ensure that the path to the environment file is always fully qualified and unambiguous.

Comment (arvindh123) [reply docker/addons/vault/scripts/vault_copy_env.sh]: Even it was fully qualified , without check it existence , we could not say that file is exists . 
Even it is relative path , if file exists is checked ,then we need to verify it is text file of env kind. (but we can skip this verification but file existence is needed)

Comment (JeffMboya) [reply docker/addons/vault/scripts/vault_copy_env.sh]: > Even it was fully qualified , without check it existence , we could not say that file is exists .
> Even it is relative path , if file exists is checked ,then we need to verify it is text file of env kind. (but we can skip this verification but file existence is needed)

Yep, that makes sense. The most important part is making sure the file is there. I think the extra check is nice to have but not absolutely necessary.

File: docker/addons/vault/scripts/vault_create_approle.sh [hunk 8]
```diff
@@ -4,39 +4,62 @@
...
+        --env-file)
+            if [[ -z "${2:-}" ]]; then
+                echo "Error: --env-file requires a non-empty option argument."
+                exit 1
+            fi
+            env_file="$2"
+            if [[ ! -f "$env_file" ]]; then
+                echo "Error: .env file not found at $env_file"
+                exit 1
+            fi
```
Comment (JackMangBana) [inline docker/addons/vault/scripts/vault_create_approle.sh:25]: @JeffMboya 
add --skil-enable-app-role

File: docker/addons/vault/scripts/vault_copy_certs.sh [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+                exit 1
+            fi
+            env_file="$2"
+            if [[ ! -f "$env_file" ]]; then
+                echo "Error: .env file not found at $env_file"
+                exit 1
+            fi
+            shift
+            ;;
+        --certs-copy-path)
```
Comment (JackMangBana) [inline docker/addons/vault/scripts/vault_copy_certs.sh:29]: @JeffMboya
This command is not working 
```bash
docker/addons/vault/scripts/vault_copy_env.sh --env-file docker/.env --certs-copy-path ./.
```
Get error 
```log 
Unknown parameter passed: --certs-copy-path
```

**Review by JackMangBana — CHANGES_REQUESTED** [review]

File: docker/addons/vault/scripts/vault_copy_certs.sh (hunk 9 above)
Comment (arvindh123) [inline docker/addons/vault/scripts/vault_copy_certs.sh:29]: For the below command i get error
```bash
docker/addons/vault/scripts/vault_copy_env.sh --env-file docker/.env --certs-copy-path ~/mainflux/
```

Error Message
```log 
Unknown parameter passed: --certs-copy-path
```

File: docker/addons/vault/scripts/vault_create_approle.sh (hunk 8 above)
Comment (arvindh123) [inline docker/addons/vault/scripts/vault_create_approle.sh:25]: Please add  flage`--skip-enable-app-role` to set  `SKIP_ENABLE_APP_ROLE`

File: docker/addons/vault/scripts/vault_copy_certs.sh [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    echo "$scriptdir/data/${MG_VAULT_PKI_INT_FILE_NAME}.key file not available"
+fi
+
+if [ -e "$scriptdir/data/${MG_VAULT_PKI_INT_FILE_NAME}_bundle.crt" ]; then
+    cp -v "$scriptdir/data/${MG_VAULT_PKI_INT_FILE_NAME}_bundle.crt" "${certs_copy_path}ca.crt"
+else
+    echo "$scriptdir/data/${MG_VAULT_PKI_INT_FILE_NAME}_bundle.crt file not available"
+fi
+
+exit 0
```
Comment (arvindh123) [inline docker/addons/vault/scripts/vault_copy_certs.sh]: add new line here

**Review by arvindh123 — CHANGES_REQUESTED** [review]

File: docker/addons/vault/scripts/vault_copy_certs.sh (hunk 9 above)
Comment (JeffMboya) [reply docker/addons/vault/scripts/vault_copy_certs.sh:29]: @arvindh123 I am not able to recreate this error

``` 
$docker/addons/vault/scripts/vault_copy_certs.sh --env-file docker/.env --certs-copy-path certs/pki     
Copying certificate files to certs/pki
'/home/jeff/Documents/magistrala/docker/addons/vault/scripts/data/localhost.crt' -> 'certs/pkimagistrala-server.crt'
'/home/jeff/Documents/magistrala/docker/addons/vault/scripts/data/localhost.key' -> 'certs/pkimagistrala-server.key'
'/home/jeff/Documents/magistrala/docker/addons/vault/scripts/data/mg_int.key' -> 'certs/pkica.key'
'/home/jeff/Documents/magistrala/docker/addons/vault/scripts/data/mg_int_bundle.crt' -> 'certs/pkica.crt'
```

Comment (JeffMboya) [reply docker/addons/vault/scripts/vault_copy_certs.sh:29]: @arvindh123 `vault_copy_env.sh` only accepts `--env-file` argument
To pass both `--env-file` and `--certs-copy-path` argument, run `vault_copy_certs.sh`

**Review by arvindh123 — APPROVED** [review]

**Review by arvindh123 — DISMISSED** [review]

File: docker/addons/vault/README.md [hunk 11]
```diff
@@ -45,22 +45,22 @@ When the Vault service is started, some initialization steps need to be done to
...
-Unseal Key 1: Ay0YZecYJ2HVtNtXfPootXK5LtF+JZoDmBb7IbbYdLBI
-Unseal Key 2: P6hb7x2cglv0p61jdLyNE3+d44cJUOFaDt9jHFDfr8Df
-Unseal Key 3: zSBfDHzUiWoOzXKY1pnnBqKO8UD2MDLuy8DNTxNtEBFy
-Unseal Key 4: 5oJuDDuMI0I8snaw/n4VLNpvndvvKi6JlkgOxuWXqMSz
-Unseal Key 5: ZhsUkk2tXBYEcWgz4WUCHH9rocoW6qZoiARWlkE5Epi5
+Unseal Key 1: `<UNSEAL_KEY_1>`
+Unseal Key 2: `<UNSEAL_KEY_2>`
+Unseal Key 3: `<UNSEAL_KEY_3>`
+Unseal Key 4: `<UNSEAL_KEY_4>`
+Unseal Key 5: `<UNSEAL_KEY_5>`
```
Comment (dborovcanin) [inline docker/addons/vault/README.md]: Revert this change. While placeholders are self-explanatory, this conflicts with the `example` premise (i.e. this is not an example anymore and tells nothing about how keys look like).

**Review by dborovcanin — APPROVED** [review]
